effekt/shared/src/main/scala/effekt/core/Transformer.scala

```diff
@@ -169,25 +169,114 @@ object Transformer extends Phase[Typechecked, CoreTransformed] {
   def transformBox(tree: source.Term)(implicit C: Context): Pure =
     Box(transformAsBlock(tree), transform(Context.inferredCapture(tree)))
 
+  /**
+   * Transforms the soruce to a function (expecting to be called using [[core.Stmt.App]] or an interface.
+   */
   def transformAsBlock(tree: source.Term)(using Context): Block = tree match {
-    case v: source.Var => v.definition match {
-      case sym: ValueSymbol => transformUnbox(tree)
-      case sym: BlockSymbol => BlockVar(sym)
-    }
-    case s @ source.Select(receiver, selector) =>
-      Member(transformAsBlock(receiver), s.definition, transform(Context.inferredBlockTypeOf(tree)))
+    case v: source.Var =>
+      val sym = v.definition
+      Context.blockTypeOf(sym) match {
+        case _: BlockType.FunctionType => transformAsControlBlock(tree)
+        case _: BlockType.InterfaceType => transformAsObject(tree)
+      }
+    case _: source.BlockLiteral => transformAsControlBlock(tree)
+    case _: source.New => transformAsObject(tree)
 
-    case s @ source.New(impl) =>
+    case s @ source.Select(receiver, id) =>
+      Member(transformAsObject(receiver), s.definition, transform(Context.inferredBlockTypeOf(tree)))
+
+    case source.Unbox(b) =>
+      Unbox(transformAsPure(b))
+
+    case _ =>
+      transformUnbox(tree)
+  }
+  /**
+   * Transforms the source to an interface block
+   */
+  def transformAsObject(tree: source.Term)(using Context): Block = tree match {
+    case v: source.Var =>
+      val sym = v.definition
+      val tpe = Context.blockTypeOf(sym)
+      tpe match {
+        case _: BlockType.FunctionType =>
+          Context.panic(s"Using function symbol ${sym} but an object was expected.")
+        case _: BlockType.InterfaceType => BlockVar(sym.asInstanceOf)
+      }
```

Comment:
Do we need this check here?

Reply:
No, this should already be catched by Typer (I think?). It is mostly there because I re-used the structure of the earlier definitions...^^'

Reply:
I removed it for simplicity.